### ISSUE TITLE
systemd-logind: improve algorithm we use to check whether hibernation is available to deal with multiple swap partitions

### ISSUE THREAD

**Issue by morfikov** [systemd/systemd#2528]
I'm trying to figure out how the `*IgnoreInhibited` options work . In the `logind.conf` manual they say that when all of the following options are set to `ignore`, systemd won't handle the keys:

```
HandlePowerKey=ignore
HandleSuspendKey=ignore
HandleHibernateKey=ignore
HandleLidSwitch=ignore
HandleLidSwitchDocked=ignore
```

So I've checked that, and I restarted the `systemd-logind.service` after that change. Now when I press, for instance, the `poweroff` button, nothing happens, which is expected. This also means that no other application holds the key, there's no additional messages logged in the journal when the key is pressed.

Now try to restore the actions that I had before, which was:

```
HandlePowerKey=hibernate
HandleSuspendKey=suspend
HandleHibernateKey=hibernate
HandleLidSwitch=suspend
HandleLidSwitchDocked=suspend
```

Restart the `systemd-logind.service`, and in the log I can see the following messages:

```
systemd-logind[12687]: Watching system buttons on /dev/input/event3 (Power Button)
systemd-logind[12687]: Watching system buttons on /dev/input/event4 (Video Bus)
systemd-logind[12687]: Watching system buttons on /dev/input/event2 (Power Button)
systemd-logind[12687]: Watching system buttons on /dev/input/event1 (Lid Switch)
systemd-logind[12687]: Watching system buttons on /dev/input/event12 (HP WMI hotkeys)
```

When the `poweroff` key is now pressed, the machine turns off.

Now to the point. When I list the inhibitors, I get the following:

```
# systemd-inhibit --list --mode=block
0 inhibitors listed.

# systemd-inhibit --list --mode=delay
     Who: Screen Locker (UID 1000/morfik, PID 2143/light-locker)
    What: sleep
     Why: Lock the screen on suspend/resume
    Mode: delay

1 inhibitors listed.
```

So if I'm right, system should turn off without problems, right? But what would happen when I started firefox and pressed the `poweroff` key? Nothing, or actually in the log I get the message:

```
systemd-logind[12687]: Power key pressed.
systemd-logind[12687]: Requested operation not supported, ignoring.
```

I'm not sure whether is related to the inhibitor lock mechanism, but when I close the browser, the key works as expected, so something is blocked here.  The above commands (`systemd-inhibit`) return the same output whether firefox is started or not.

I tried to change some other options found in the `login.conf` file:

```
PowerKeyIgnoreInhibited=yes
SuspendKeyIgnoreInhibited=yes
HibernateKeyIgnoreInhibited=yes
LidSwitchIgnoreInhibited=yes
```

If I understood the options well, they should allow pressing the buttons even if some applications are listed under `systemd-inhibit --list`. I though maybe that would help, so I've set the options and restarted the `systemd-logind.service` once more. But this actually does nothing. I still get the message when firefox is started:

```
systemd-logind[17193]: Power key pressed.
systemd-logind[17193]: Requested operation not supported, ignoring.
```

So what's wrong? Why the machine is unable to power off?

**Comment by arvidjaar** [comment]
> when I close the browser, the key works as expected

Could be that you have not enough swap space to save dirty memory. Firefox needs quite a bit.

**Comment by morfikov** [comment]
So, for example, if I had 2G of RAM, and when the key is pressed the system utilizes around 1.1G (0 of 2G in swap), could this still be the issue? What about the sysctl parameters:

```
# sysctl -a | grep dirty
vm.dirty_background_bytes = 0
vm.dirty_background_ratio = 3
vm.dirty_bytes = 0
vm.dirty_expire_centisecs = 400
vm.dirty_ratio = 5
vm.dirty_writeback_centisecs = 1000
vm.dirtytime_expire_seconds = 43200
```

**Comment by poettering** [comment]
So, let me get this right, you configure your powerkey to result in hibernation, but hibernation doesn't work? Does it work if you request it via "systemctl hibernate"?

What does /sys/power/state and /sys/power/disk contain if you cat them?

**Comment by morfikov** [comment]
No, it doesn't work. I get the following message:

```
$ systemctl hibernate
Failed to hibernate system via logind: Sleep verb not supported
```

The files contain this:

```
$ cat /sys/power/state
freeze mem disk

$ cat /sys/power/disk
[platform] shutdown reboot suspend
```

**Comment by poettering** [comment]
yeah, seems logind came to the conclusion that your setup doesn't support hibernation. Consider turning on debug mode in logind (by adding Environment=SYSTEMD_LOG_LEVEL=debug to systemd-logind.service -– `systemctl --edit` is your friend), it will then give a longer explanation in the logs.

**Comment by morfikov** [comment]
@poettering , hmm, now I got this message:

```
systemd-logind[19891]: Hibernation is impossible, Active(anon)=603064 kB, size=524284 kB, used=90732 kB, threshold=98%
```

So I checked `swapon`:

```
# swapon
NAME       TYPE      SIZE  USED  PRIO
/dev/zram0 partition 512M 86.5M 16383
/dev/dm-4  partition   2G    0B    10
```

The /dev/dm-4  partition is my encrypted SWAP. I have also the following line in `/etc/initramfs-tools/conf.d/resume` file:

```
RESUME=/dev/mapper/debian_laptop-swap
```

This is the same partition as /dev/dm-4 . But the numbers printed by `systemd-logind` point to the `zram` device and the question is why and is there a way to force `systemd-logind` to use the other SWAP?

**Comment by arvidjaar** [comment]
logind always assumes the very first swap device. This is probably wrong, kernel will use resume device (the one you set using resume= kernel parameter or by writing into /sys/power/resume) so logind should take device from /sys/power/resume and verify that it is swap device.

**Comment by poettering** [comment]
I guess this can be closed now that #13089 and 411ae92b407bd7b4549b205ad754bcd0e3dfd81f have been merged.
